blocks/monitoring/evaluators.py

```diff
@@ -172,7 +196,20 @@ class DatasetEvaluator(object):
 
     """
     def __init__(self, variables, updates=None):
-        self.buffer_ = AggregationBuffer(variables)
+        theano_variables = []
+        non_theano_variables = []
+        for variable in variables:
+            if isinstance(variable, MonitoredQuantity):
+                non_theano_variables += [variable]
+            else:
+                theano_variables += [variable]
+        self.theano_variables = theano_variables
+        self.non_theano_variables = non_theano_variables
+        variable_names = [v.name for v in theano_variables + non_theano_variables]
```

Comment:
I guess you are missing `set` here, list of names will always be just as long as the list of variables.